File: src/core/codex.ts
/**
 * El codex: el registro de lo que descubriste.
 *
 * Hasta acá el juego generaba una variedad enorme y no quedaba ningún rastro de
 * ella. Cada criatura nueva pisaba a la anterior y los linajes, las formas y las
 * rarezas que hubieras visto se perdían apenas cerrabas.
 *
 * El codex existe para eso: guardar qué apareció alguna vez y, sobre todo,
 * mostrar qué falta. Sin la parte de "qué falta" no es una colección, es un
 * historial.
 *
 * Módulo puro: recibe un genoma y una forma, devuelve un codex nuevo. No muta
 * nada ni sabe de dónde salieron los datos.
 */

import { COLLECTIBLE_FORMS, type Form, formName } from "./evolution.ts";
import { LINEAGES, decodeGenome, lineageName } from "./genome.ts";
import { TRAIT_CATALOG, detectTraits } from "./traits.ts";

export interface Codex {
  /** Índices de linaje descubiertos, ordenados. */
  linajes: number[];
  /** Formas evolutivas alcanzadas alguna vez. */
  formas: Form[];
  /** Ids de rarezas encontradas. */
  rarezas: string[];
  /** Cuántas criaturas se registraron en total, incluidas las repetidas. */
  totalRegistradas: number;
}

export function codexInicial(): Codex {
  return { linajes: [], formas: [], rarezas: [], totalRegistradas: 0 };
}

export type TipoDescubrimiento = "linaje" | "forma" | "rareza";

export interface Descubrimiento {
  tipo: TipoDescubrimiento;
  /** Identificador interno, para la UI. */
  id: string;
  /** Nombre legible, ya redactado. */
  nombre: string;
}

/**
 * Registra una criatura en el codex.
 *
 * Devuelve además qué fue novedad. Es información que la UI necesita: descubrir
 * algo por primera vez tiene que sentirse distinto a volver a verlo, y sin esto
 * habría que comparar el codex viejo contra el nuevo desde afuera.
 */
export function registrar(
  codex: Codex,
  seed: bigint,
  forma: Form,
): { codex: Codex; nuevos: Descubrimiento[] } {
  const genes = decodeGenome(seed);
  const nuevos: Descubrimiento[] = [];

  const linajes = [...codex.linajes];
  if (!linajes.includes(genes.lineage)) {
    linajes.push(genes.lineage);
    nuevos.push({
      tipo: "linaje",
      id: String(genes.lineage),
      nombre: lineageName(genes),
    });
  }

  const formas = [...codex.formas];
  // "indefinida" no es un descubrimiento: es la ausencia de uno.
  if (forma !== "indefinida" && !formas.includes(forma)) {
    formas.push(forma);
    nuevos.push({ tipo: "forma", id: forma, nombre: formName(forma) });
  }

  const rarezas = [...codex.rarezas];
  for (const trait of detectTraits(seed)) {
    if (!rarezas.includes(trait.id)) {
      rarezas.push(trait.id);
      nuevos.push({ tipo: "rareza", id: trait.id, nombre: trait.name });
    }
  }

  return {
    codex: {
      // Ordenados para que el guardado sea estable: si no, dos partidas
      // equivalentes producen JSON distinto y cualquier comparación miente.
      linajes: linajes.sort((a, b) => a - b),
      formas: formas.slice().sort(),
      rarezas: rarezas.slice().sort(),
      totalRegistradas: codex.totalRegistradas + 1,
    },
    nuevos,
  };
}

export interface Avance {
  vistos: number;
  total: number;
}

export interface ProgresoCodex {
  linajes: Avance;
  formas: Avance;
  rarezas: Avance;
  /** Porcentaje global de completitud, 0-100. */
  porcentaje: number;
}

/**
 * Cuánto llevás descubierto.
 *
 * Las dos rarezas legendarias entran en el total aunque sean prácticamente
 * inalcanzables — Pangrama es 1 en 880.000. Es a propósito: un codex que se
 * completa del todo deja de dar motivo para seguir mirando.
 */
export function progresoCodex(codex: Codex): ProgresoCodex {
  const linajes: Avance = { vistos: codex.linajes.length, total: LINEAGES.length };
  const formas: Avance = { vistos: codex.formas.length, total: COLLECTIBLE_FORMS.length };
  const rarezas: Avance = { vistos: codex.rarezas.length, total: TRAIT_CATALOG.length };

  const vistos = linajes.vistos + formas.vistos + rarezas.vistos;
  const total = linajes.total + formas.total + rarezas.total;

  return { linajes, formas, rarezas, porcentaje: Math.round((vistos / total) * 100) };
}

/** ¿Ya se había visto este linaje? Para pintar la ficha sin recalcular todo. */
export function conoceLinaje(codex: Codex, lineage: number): boolean {
  return codex.linajes.includes(lineage);
}

export function conoceForma(codex: Codex, forma: Form): boolean {
  return codex.formas.includes(forma);
}

export function conoceRareza(codex: Codex, id: string): boolean {
  return codex.rarezas.includes(id);
}

File: gdext/src/codex.cpp
#include "codex.h"

#include "js_math.h"

#include <algorithm>

namespace petbits {

std::string_view nombreTipo(TipoDescubrimiento t) {
    switch (t) {
        case TipoDescubrimiento::Linaje: return "linaje";
        case TipoDescubrimiento::Forma:  return "forma";
        case TipoDescubrimiento::Rareza: return "rareza";
    }
    return "?";
}

Registro registrar(const Codex& codex, Seed seed, Form forma) {
    const Genes genes = decodeGenome(seed);

    Registro r;
    r.codex = codex;

    const int linaje = static_cast<int>(genes.lineage);
    if (std::find(r.codex.linajes.begin(), r.codex.linajes.end(), linaje) ==
        r.codex.linajes.end()) {
        r.codex.linajes.push_back(linaje);
        r.nuevos.push_back({TipoDescubrimiento::Linaje, std::to_string(linaje),
                            std::string(lineageName(genes))});
    }

    // "Indefinida" no es un descubrimiento: es la ausencia de uno.
    if (forma != Form::Indefinida &&
        std::find(r.codex.formas.begin(), r.codex.formas.end(), forma) ==
            r.codex.formas.end()) {
        r.codex.formas.push_back(forma);
        r.nuevos.push_back(
            {TipoDescubrimiento::Forma, std::string(formId(forma)), std::string(formName(forma))});
    }

    for (const Trait& trait : detectTraits(seed)) {
        const std::string id(trait.id);
        if (std::find(r.codex.rarezas.begin(), r.codex.rarezas.end(), id) ==
            r.codex.rarezas.end()) {
            r.codex.rarezas.push_back(id);
            r.nuevos.push_back({TipoDescubrimiento::Rareza, id, std::string(trait.name)});
        }
    }

    // Ordenados para que el guardado sea estable: si no, dos partidas
    // equivalentes producen JSON distinto y cualquier comparación miente.
    //
    // Los tres órdenes son distintos entre sí y ninguno es el obvio:
    std::sort(r.codex.linajes.begin(), r.codex.linajes.end());
    // Las formas van por su ID EN TEXTO y no por el valor del enum. El enum
    // arranca en Indefinida, Petreo, Vaporoso…; alfabéticamente el primero es
    // "coloso". Ordenar por enum daría otro archivo, y el validador de la web no
    // se quejaría — solo dejaría de coincidir con lo que escribe el navegador.
    std::sort(r.codex.formas.begin(), r.codex.formas.end(),
              [](Form a, Form b) { return formId(a) < formId(b); });
    std::sort(r.codex.rarezas.begin(), r.codex.rarezas.end());

    r.codex.totalRegistradas = codex.totalRegistradas + 1;
    return r;
}

ProgresoCodex progresoCodex(const Codex& codex) {
    ProgresoCodex p;
    p.linajes = {static_cast<int>(codex.linajes.size()), static_cast<int>(LINEAGES.size())};
    p.formas = {static_cast<int>(codex.formas.size()),
                static_cast<int>(formasColeccionables().size())};
    p.rarezas = {static_cast<int>(codex.rarezas.size()),
                 static_cast<int>(TRAIT_CATALOG.size())};

    const int vistos = p.linajes.vistos + p.formas.vistos + p.rarezas.vistos;
    const int total = p.linajes.total + p.formas.total + p.rarezas.total;

    // jsRound y no std::round: el Math.round de JavaScript redondea el .5 hacia
    // arriba y std::round lo aleja del cero, que difieren en los negativos. Acá
    // nunca hay negativos, pero usar el helper del proyecto es más barato que
    // dejar escrito que "en este caso da igual" y que alguien lo mueva.
    p.porcentaje = static_cast<int>(
        jsRound(static_cast<double>(vistos) / static_cast<double>(total) * 100.0));
    return p;
}

bool conoceLinaje(const Codex& codex, int lineage) {
    return std::find(codex.linajes.begin(), codex.linajes.end(), lineage) != codex.linajes.end();
}

bool conoceForma(const Codex& codex, Form forma) {
    return std::find(codex.formas.begin(), codex.formas.end(), forma) != codex.formas.end();
}

bool conoceRareza(const Codex& codex, std::string_view id) {
    return std::find(codex.rarezas.begin(), codex.rarezas.end(), id) != codex.rarezas.end();
}

} // namespace petbits
